helix-gateway/src/main/java/org/apache/helix/gateway/service/GatewayServiceEvent.java

```diff
@@ -0,0 +1,95 @@
+package org.apache.helix.gateway.service;
+
+import java.util.Map;
+import org.apache.helix.gateway.constant.GatewayServiceEventType;
+
+
+public  class GatewayServiceEvent {
+  // event type
+  GatewayServiceEventType _eventType;
+  // event data
+  String _clusterName;
+  String _instanceName;
+  Map<String, Map<String, String>> _shardStateMap;
+
+  StateTransitionResult _StateTransitionResult;
+
+  public class StateTransitionResult {
+    String stateTransitionId;
+    String stateTransitionStatus;
+    String shardState;
+
+    public StateTransitionResult(String stateTransitionId, String stateTransitionStatus, String shardState) {
+      this.stateTransitionId = stateTransitionId;
+      this.stateTransitionStatus = stateTransitionStatus;
+      this.shardState = shardState;
+    }
+  }
+
+  private GatewayServiceEvent(GatewayServiceEventType eventType, String clusterName, String instanceName) {
+    this._eventType = eventType;
+    this._clusterName = clusterName;
+    this._instanceName = instanceName;
+  }
+
+  private GatewayServiceEvent(GatewayServiceEventType eventType, String clusterName, String instanceName,  Map<String, Map<String, String>> shardStateMap) {
+    this._eventType = eventType;
+    this._clusterName = clusterName;
+    this._instanceName = instanceName;
+    this._shardStateMap = shardStateMap;
+  }
+
+  private GatewayServiceEvent(GatewayServiceEventType eventType, String clusterName, String instanceName, String stateTransitionId,
+      String stateTransitionStatus, String shardState) {
+    this._eventType = eventType;
+    this._clusterName = clusterName;
+    this._instanceName = instanceName;
+    this._StateTransitionResult = new StateTransitionResult(stateTransitionId, stateTransitionStatus, shardState);
+  }
+
```

Comment:
If we have a builder, no need to have so many constructor.

Also even as public constructor, we should call "this(xxxxxxx)" to reduce redundant code.